web3/utils/abi.py

```diff
@@ -1,21 +1,530 @@
+        cast(ABICallable, function_abi)
+        for function_abi in contract_abi
+        if check_if_arguments_can_be_encoded(
+            function_abi, *args, abi_codec=abi_codec, **kwargs
+        )
+    ]
+
+
+def _mismatched_abi_error_diagnosis(
+    function_identifier: FunctionIdentifier,
+    matching_function_signatures: Sequence[str],
+    arg_count_matches: int,
+    encoding_matches: int,
+    *args: Optional[Sequence[Any]],
+    **kwargs: Optional[Dict[str, Any]],
+) -> str:
+    """
+    Raise a ``MismatchedABI`` when a function ABI lookup results in an error.
+
+    An error may result from multiple functions matching the provided signature and
+    arguments or no functions are identified.
+    """
+    diagnosis = "\n"
+    if arg_count_matches == 0:
+        diagnosis += "Function invocation failed due to improper number of arguments."
+    elif encoding_matches == 0:
+        diagnosis += "Function invocation failed due to no matching argument types."
+    elif encoding_matches > 1:
+        diagnosis += (
+            "Ambiguous argument encoding. "
+            "Provided arguments can be encoded to multiple functions "
+            "matching this call."
+        )
+
+    collapsed_args = _extract_argument_types(*args)
+    collapsed_kwargs = dict(
+        {(k, _extract_argument_types([v])) for k, v in kwargs.items()}
+    )
+
+    return (
+        f"\nCould not identify the intended function with name "
+        f"`{function_identifier}`, positional arguments with type(s) "
+        f"`({collapsed_args})` and keyword arguments with type(s) "
+        f"`{collapsed_kwargs}`."
+        f"\nFound {len(matching_function_signatures)} function(s) with the name "
+        f"`{function_identifier}`: {matching_function_signatures}{diagnosis}"
+    )
+
+
+def _extract_argument_types(*args: Sequence[Any]) -> str:
+    """
+    Takes a list of arguments and returns a string representation of the argument types,
+    appropriately collapsing `tuple` types into the respective nested types.
+    """
+    collapsed_args = []
+
+    for arg in args:
+        if is_list_like(arg):
+            collapsed_nested = []
+            for nested in arg:
+                if is_list_like(nested):
+                    collapsed_nested.append(f"({_extract_argument_types(nested)})")
+                else:
+                    collapsed_nested.append(_get_argument_readable_type(nested))
+            collapsed_args.append(",".join(collapsed_nested))
+        else:
+            collapsed_args.append(_get_argument_readable_type(arg))
+
+    return ",".join(collapsed_args)
+
+
+def _get_argument_readable_type(arg: Any) -> str:
+    """
+    Returns the class name of the argument, or `address` if the argument is an address.
+    """
+    if is_checksum_address(arg) or is_binary_address(arg):
+        return "address"
+
+    return arg.__class__.__name__
+
+
+def get_abi_element_info(
+    abi: ABI,
+    function_identifier: FunctionIdentifier,
```

Comment:
Somewhat related to [this thread](https://github.com/ethereum/web3.py/pull/3408/files#r1695774197), should this type reflect an abi element, rather than just a function identifier? Otherwise the name should be something closer to `get_abi_callable_info()`.

Reply:
The identifier has been historically used as a way to distinguish `fallback` and `receive` types, but the others are all used with just a `str` name. I think this should be named `element_identifier`, and the type could potentially change too. Should I update both?